codex/merkletree/codex/codex.nim

```diff
@@ -145,16 +145,18 @@ func compress*(
   mhash.coder(@x & @y & @[ key.byte ], digest)
   success digest
 
-func init*(
+proc init*(
   _: type CodexTree,
   mcodec: MultiCodec = Sha256HashCodec,
-  leaves: openArray[ByteHash]): ?!CodexTree =
+  leaves: seq[ByteHash]): Future[?!CodexTree] {.async.} =
 
   if leaves.len == 0:
     return failure "Empty leaves"
 
+  without mhash =? mcodec.mhash(), error:
+    return failure error
+
   let
-    mhash = ? mcodec.mhash()
```

Comment:
> we could simply preserve this behavior and propagate the exception using a Chronos raises clause

That would actually change the behaviour. `= ?` will either assign a value to `mhash` or it will return a `Result` with an error to the closure. By raising an exception, and specifying that exception type in `{.async: (raises: [...]).}`, we're not returning a Result, but raising an exception. 

Imo, the proposed change here does not alter the behaviour.